app/src/main/java/app/revanced/manager/patcher/worker/PatcherProgressManager.kt

```diff
@@ -9,117 +9,92 @@ import kotlinx.collections.immutable.persistentListOf
 import kotlinx.collections.immutable.toImmutableList
 import kotlinx.coroutines.flow.StateFlow
 
+enum class StepCategory(@StringRes val displayName: Int) {
+    PREPARING(R.string.patcher_step_group_preparing),
+    PATCHING(R.string.patcher_step_group_patching),
+    SAVING(R.string.patcher_step_group_saving)
+}
+
 enum class State {
-    WAITING, COMPLETED, FAILED
+    WAITING, RUNNING, FAILED, COMPLETED
 }
 
-class SubStep(
+data class Step(
     val name: String,
+    val category: StepCategory,
     val state: State = State.WAITING,
     val message: String? = null,
     val progress: StateFlow<Pair<Float, Float>?>? = null
 )
 
-class Step(
-    @StringRes val name: Int,
-    val subSteps: ImmutableList<SubStep>,
-    val state: State = State.WAITING
-)
-
 class PatcherProgressManager(
     context: Context,
     selectedPatches: List<String>,
     selectedApp: SelectedApp,
-    downloadProgress: StateFlow<Pair<Float, Float>?>
+    progress: StateFlow<Pair<Float, Float>?>
 ) {
-    val steps = generateSteps(context, selectedPatches, selectedApp, downloadProgress)
-    private var currentStep: StepKey? = StepKey(0, 0)
+    var steps: ImmutableList<Step> = persistentListOf()
+        private set
 
-    private fun update(key: StepKey, state: State, message: String? = null) {
-        val isLastSubStep: Boolean
-        steps[key.step] = steps[key.step].let { step ->
-            isLastSubStep = key.substep == step.subSteps.lastIndex
+    private var currentStepIndex: Int = 0
 
-            val newStepState = when {
-                // This step failed because one of its sub-steps failed.
-                state == State.FAILED -> State.FAILED
-                // All sub-steps succeeded.
-                state == State.COMPLETED && isLastSubStep -> State.COMPLETED
-                // Keep the old status.
-                else -> step.state
-            }
+    init {
+        steps = generateSteps(
+            context,
+            selectedPatches,
+            selectedApp,
+            progress
+        )
+    }
 
-            Step(step.name, step.subSteps.mapIndexed { index, subStep ->
-                if (index != key.substep) subStep else SubStep(subStep.name, state, message)
-            }.toImmutableList(), newStepState)
-        }
+    fun updateProgress(state: State, message: String? = null) {
+        steps = steps.toMutableList().apply {
+            this[currentStepIndex] = this[currentStepIndex].copy(state = state, message = message)
```

Comment:
I think this code would be cleaner if you switch to a `PersistentList` instead of an `ImmutableList`